glob6.1
.\"
.\" Copyright (c) 2006-2011
.\"	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
.\"	All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	@(#)$Id$
.\"
.\"	Derived from:
.\"		- Sixth Edition UNIX	/usr/man/man1/sh.1
.\"		- Sixth Edition UNIX	/usr/man/man8/glob.8
.\"
.\" Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code and documentation must retain the above
.\"    copyright notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\" 4. Neither the name of Caldera International, Inc. nor the names of other
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
.\" INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH GLOB6 1 "@OSH_DATE@" "@OSH_VERSION@" "General Commands"
.SH NAME
glob6 \- global command (file name generation)
.SH SYNOPSIS
.B glob6
\fIcommand\fR [\fIarg ...\fR]
.SH DESCRIPTION
.I Glob6
is a port of the global command from Sixth Edition UNIX.
It is passed an argument list by
.IR sh6 (1)
which contains one or more unquoted
.I "pattern characters"
(`*', `?', `[')
and performs the actions described below
on behalf of the shell.
.PP
.I Glob6
attempts to generate file-name arguments which match
the given pattern arguments.
The name of the specified
.I command
may also be given as a pattern.
The meaning of each pattern character is as follows:
.IP o 4
The `*' character in a pattern matches any string of
characters in a file name (including the null string).
.IP o
The `?' character in a pattern matches any single character
in a file name.
.IP o
The `[...]' brackets in a pattern specifies a class of characters
which matches any single file-name character in the class.
Within the brackets,
each character is taken to be a member of the class.
A pair of characters separated by an unquoted `\-' specifies
the class as a range which matches each character lexically
between the first and second member of the pair, inclusive.
A `\-' matches itself when quoted or when first or last
in the class.
.PP
Any other character in a pattern matches itself in a file name.
.PP
Notice that the `.' character at the beginning of a file name,
or immediately following a `/',
is always special in that it must be matched explicitly.
The same is true of the `/' character itself.
.PP
If the pattern contains no `/' characters,
the current directory is always used.
Otherwise,
the specified directory is the one obtained by taking the pattern
up to the last `/' before the first unquoted `*', `?', or `['.
The matching process matches the remainder of the pattern
after this `/' against the files in the specified directory.
.PP
If the argument contains no unquoted pattern characters,
.I glob6
uses it as is.
Otherwise,
.I glob6
searches for file names in the current (or specified) directory
which match the pattern.
It then sorts them in ascending ASCII order,
and the new sequence of arguments replaces the given pattern.
The same process is carried out for each of the given arguments;
the resulting lists are
.I not
merged.
Finally,
.I glob6
attempts to execute the command
with the resulting argument list.
.SH "EXIT STATUS"
If
.I glob6
detects an error,
it prints an appropriate diagnostic
and exits with a non-zero status.
Otherwise,
the exit status is that
of the executed command.
.SH ENVIRONMENT
Notice that the concept of `user environment'
was not defined in Sixth Edition UNIX.
Thus,
use of the following environment variables
by this port of the global command
is an enhancement:
.TP
.B EXECSHELL
If set to a non-empty string,
the value of this variable is taken as the
path name of the shell which is invoked to
execute the specified command when it does not
begin with the proper magic number
or a `#!shell' sequence.
.TP
.B PATH
If set to a non-empty string,
the value of this variable is taken as the
sequence of directories which is used to
search for the specified command.
Notice that the
global command from Sixth Edition UNIX
always used the equivalent of `.:/bin:/usr/bin',
not PATH.
.SH "SEE ALSO"
sh6(1)
.PP
Osh home page:
http://v6shell.org/
.SH HISTORY
A
.I glob
command
appeared as
.I /etc/glob
in First Edition UNIX.
.SH LICENSE
See either the LICENSE file which is distributed with
.I osh
or
http://v6shell.org/license/
for full details.
.SH COPYRIGHT
.nf
Copyright (c) 2006-2011
	Jeffrey Allen Neitzel.  All rights reserved.

Copyright (c) 2001-2002
	Caldera International Inc.  All rights reserved.

Copyright (c) 1985, 1989, 1991, 1993
	The Regents of the University of California.  All rights reserved.
.fi


CHANGES
Changes that improve compatibility w/ the Sixth Edition (V6) UNIX
Thompson shell are marked w/ a `C:' in the details below.

-------------------------------------------------------------------------------
[osh-current]: Friday, 2010-05-14:
osh.1:
	* Added [--] to fd2 usage synopsis in "Special commands" subsection.

	* s/Mac/Mac OS X/

-------------------------------------------------------------------------------
[osh-20100430]:
osh.c:
	* Added a new, osh history-file-support feature to save the user's
	  command-line history to the user $h/.osh.history file.  Notice
	  that the history file must already exist as a writable, regular
	  file (or a link to a regular file) when the shell is invoked to
	  save the user's command-line history.  Otherwise, it won't do so.

	  An interactive shell reads each command line from its terminal and
	  writes the words of each one to the history file as a history entry
	  after performing parameter substitution and word splitting.

osh.1:
	* Documented the new, osh history-file-support feature in the new
	  "History file (+)" subsection.

-------------------------------------------------------------------------------
[osh-20100228]:
mkconfig:
	* Added a search comparison that uses which(1) to find login & newgrp
	  pathname(s).  This allows for those cases where a sysadmin may use
	  locally modified versions of these utilities to deal w/ uniquely
	  local needs.  Notice that this is in addition to how login & newgrp
	  pathname(s) were searched for before now.

osh.c:
	* Changed osh globbing so it's handled in main osh process.  This is
	  for user convenience to allow (for example):

		% mkdir some_really_long_directory_name... ; cd some*...

	  ...
	  and globbing for other special built-in commands as well.

-------------------------------------------------------------------------------
[osh-20091218]:
	* Incorporated osh-20091127-p1.patch changes (see p1: below).

INSTALL, Makefile:
	* Made "install-doc" and "install-exp" targets optional so that
	  they're no longer triggered by default "install" target(s).

mkconfig:
	* p1: Defined CONFIG_SUNOS on OpenSolaris (uname -s == SunOS).

goto.c:
	* p1: Fixed problem w/ external goto command on OpenSolaris where
	      file offset set via lseek(2) isn't shared w/ parent process,
	      causing unexpected behavior.

examples/*osh*:
	* Made changes that allow users to use osh to initialize/invoke sh6
	  by doing `[exec] osh [-i | -l] sh6'. This is most convenient way
	  to perform startup actions for sh6 without adding undesirable
	  functionality to sh6 itself.

examples/dot.osh.logout:
	* Fixed `rm -r $d' error on OpenSolaris.

-------------------------------------------------------------------------------
[osh-20091127]:
err.c:
	* Changed setmyname(argv[0]) so that it only ditches an initial
	  "-" character when a binary name is "-" plus at least 1 of any
	  character before end-of-string.  For example:

		% pwd ; ln -s osh - ; ln -s osh -o ; ln -s osh -os ; \
			ln -s osh -osh ; echo -* osh
		/tmp/v6shell
		- -o -os -osh osh

		% which - -o -os -osh osh
		/tmp/v6shell/-
		/tmp/v6shell/-o
		/tmp/v6shell/-os
		/tmp/v6shell/-osh
		/tmp/v6shell/osh

		% - file arg1 ; -o file arg1 ; -os file arg1 ; \
			-osh file arg1 ; osh file arg1
		-: file: cannot open
		o: file: cannot open
		os: file: cannot open
		osh: file: cannot open
		osh: file: cannot open

-------------------------------------------------------------------------------
[osh-20091122]:
defs.h:
	* Defined SBUFMM(m), small buffer max multiplier, for calculating
	  small buffer sizes (e.g., DOLMAX, FMTMAX, LABELMAX).

err.h:
	* Changed ERR_ALTOOLONG to ERR_E2BIG, and added FMT4S.

pexec.[hc]:
	* Changed err_pexec() function prototype.

	* Changed it to print ERR_E2BIG diagnostic when errno == E2BIG.

	* Changed exit status to 126 for such "cannot execute" conditions.

*.c:
	* Changed all err_pexec() calls to match new function prototype
	  where appropriate.

osh.c:
	* Changed main child process execute function from pexec()
	  to err_pexec() for external commands.

	* Added getmyname() program name prefix (e.g., "osh: ")
	  to osh diagnostics.

v.c:
	* Added "The Regents of the University of California" copyright notice
	  for NetBSD and OpenBSD code included in the osh project.

[fgio][dfos]*.1:
	* Added COPYRIGHT section headings, and added copyright notices.

-------------------------------------------------------------------------------
[osh-20091029]:
	* Remove rcsid.h file; stop adding rcsid strings for all
	  sources to compiled binaries.

	* Re-add example rc (init and logout) files as examples/* .

	* Add $h/.project file as PROJECT for quick project info.

Makefile:
	* Add default DOCDIR as $(PREFIX)/share/doc/$(OSH_VERSION);
	  use it for [ACILNPR]* package documentation and examples/* .

	* Change default MANDIR to $(PREFIX)/share/man/man1 .

v.c:
	* Include OSH_VERSION from Makefile and project copyright
	  notices in compiled binaries.

fd2.c, if.c, pexec.[hc], util.c:
	* Add err_pexec() error-handling wrapper around pexec() for
	  shell utilities that execute external commands.

osh.c:
	* Wrap WCOREDUMP(s) w/ #ifdef WCOREDUMP ... #endif .

osh.1:
	* Change example in "I/O redirection" subsection that could
	  confuse new users.

-------------------------------------------------------------------------------
[osh-20090527]:
	* Moved inclusion of required kernel include files (e.g., <sys/*.h>)
	  and /usr/include files from *.c to "defs.h" where appropriate.

	* Moved files:	Moved manual pages from *.1.in to *.1 .

	* New files:	err.h, err.c  (see err.[hc] description below)

Makefile:
	* Changed "-O2" to "-Os" in CFLAGS to optimize for size.

	* Changed how the manual pages are generated and installed.

README:
	* Added Debian 5.0.1 (lenny), NetBSD 5.0, OpenBSD 4.5, and
	  OpenSolaris 2008.11 to the reported system list; the osh
	  package behaves as expected on each system.

err.[hc]:

glob6.c
/*
 * glob6.c - a port of the Sixth Edition (V6) UNIX global command
 */
/*-
 * Copyright (c) 2004-2011
 *	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
 *	All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *	@(#)$Id$
 */
/*
 *	Derived from: Sixth Edition UNIX /usr/source/s1/glob.c
 */
/*-
 * Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code and documentation must retain the above
 *    copyright notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed or owned by Caldera
 *      International, Inc.
 * 4. Neither the name of Caldera International, Inc. nor the names of other
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
 * INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
 * INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "defs.h"
#include "err.h"
#include "pexec.h"

static	const char	**gavp;	/* points to current gav position     */
static	const char	**gave;	/* points to current gav end          */
static	size_t		gavtot;	/* total bytes used for all arguments */

static	const char	**gavnew(/*@only@*/ const char **);
static	char		*gcat(const char *, const char *);
static	const char	**glob1(/*@only@*/ const char **, char *, int *);
static	bool		glob2(const UChar *, const UChar *);
static	void		gsort(const char **);
/*@null@*/
static	DIR		*gopendir(const char *);

/*
 * NAME
 *	glob6 - global command (file name generation)
 *
 * SYNOPSIS
 *	glob6 command [arg ...]
 *
 * DESCRIPTION
 *	See glob6(1) the manual page for full details.
 */
int
main(int argc, char **argv)
{
	const char **gav;	/* points to generated argument vector */
	int pmc = 0;		/* pattern-match count                 */

	setmyerrexit(&ut_errexit);
	setmyname(argv[0]);
	setmypid(getpid());

	/*
	 * Set-ID execution is not supported.
	 */
	if (geteuid() != getuid() || getegid() != getgid())
		err(SH_ERR, FMT1S, ERR_SETID);

	if (argc < 2)
		err(SH_ERR, FMT1S, ERR_GARGCOUNT);

	if ((gav = malloc(GAVNEW * sizeof(char *))) == NULL)
		err(SH_ERR, FMT1S, ERR_NOMEM);

	*gav = NULL, gavp = gav;
	gave = &gav[GAVNEW - 1];
	while (*++argv != NULL)
		gav = glob1(gav, *argv, &pmc);
	gavp = NULL;

	if (pmc == 0)
		err(SH_ERR, FMT1S, ERR_NOMATCH);

	(void)err_pexec(gav[0], (char *const *)gav);
	/*NOTREACHED*/
	return SH_ERR;
}

static const char **
gavnew(const char **gav)
{
	size_t siz;
	ptrdiff_t gidx;
	const char **nav;
	static unsigned mult = 1;

	if (gavp == gave) {
		mult *= GAVMULT;
		gidx  = (ptrdiff_t)(gavp - gav);
		siz   = (size_t)((gidx + (GAVNEW * mult)) * sizeof(char *));
		if ((nav = realloc(gav, siz)) == NULL)
			err(SH_ERR, FMT1S, ERR_NOMEM);
		gav   = nav;
		gavp  = gav + gidx;
		gave  = &gav[gidx + (GAVNEW * mult) - 1];
	}
	return gav;
}

static char *
gcat(const char *src1, const char *src2)
{
	size_t siz;
	char *b, buf[PATHMAX], c, *dst;
	const char *s;

	*buf = EOS, b = buf, s = src1;
	while ((c = *s++) != EOS) {
		if (b >= &buf[PATHMAX - 1])
			err(SH_ERR, FMT1S, strerror(ENAMETOOLONG));
		if ((c &= ASCII) == EOS) {
			*b++ = SLASH;
			break;
		}
		*b++ = c;
	}

	s = src2;
	do {
		if (b >= &buf[PATHMAX])
			err(SH_ERR, FMT1S, strerror(ENAMETOOLONG));
		*b++ = c = *s++;
	} while (c != EOS);

	siz = b - buf;
	gavtot += siz;
	if (gavtot > GAVMAX)
		err(SH_ERR, FMT1S, ERR_E2BIG);
	if ((dst = malloc(siz)) == NULL) {
		err(SH_ERR, FMT1S, ERR_NOMEM);
		/*NOTREACHED*/
	}

	(void)memcpy(dst, buf, siz);
	return dst;
}

static const char **
glob1(const char **gav, char *as, int *pmc)
{
	DIR *dirp;
	struct dirent *entry;
	ptrdiff_t gidx;
	const char *ds;
	char *ps;

	ds = ps = as;
	while (*ps != ASTERISK && *ps != QUESTION && *ps != LBRACKET)
		if (*ps++ == EOS) {
			gav = gavnew(gav);
			*gavp++ = gcat(as, "");
			*gavp = NULL;
			return gav;
		}
	if (strlen(as) >= PATHMAX)
		err(SH_ERR, FMT1S, ERR_PATTOOLONG);
	for (;;) {
		if (ps == ds) {
			ds = "";
			break;
		}
		if ((*--ps & ASCII) == SLASH) {
			*ps = EOS;
			if (ds == ps)
				ds = "/";
			*ps++ = (char)QUOTE;
			break;
		}
	}
	if ((dirp = gopendir(*ds != EOS ? ds : ".")) == NULL) {
		err(SH_ERR, FMT1S, ERR_NODIR);
		/*NOTREACHED*/
	}
	gidx = (ptrdiff_t)(gavp - gav);
	while ((entry = readdir(dirp)) != NULL) {
		if (entry->d_name[0] == DOT && (*ps & ASCII) != DOT)
			continue;
		if (glob2(UCPTR(entry->d_name), UCPTR(ps))) {
			gav = gavnew(gav);
			*gavp++ = gcat(ds, entry->d_name);
			(*pmc)++;
		}
	}
	(void)closedir(dirp);
	gsort(gav + gidx);
	*gavp = NULL;
	return gav;
}

static bool
glob2(const UChar *ename, const UChar *pattern)
{
	int cok, rok;		/* `[...]' - cok (class), rok (range) */
	UChar c, ec, pc;
	const UChar *e, *n, *p;

	e = ename;
	p = pattern;
	if ((ec = *e++) != EOS)
		if ((ec &= ASCII) == EOS)
			ec = (UChar)QUOTE;

	switch (pc = *p++) {
	case EOS:
		return ec == EOS;

	case ASTERISK:
		/*
		 * Ignore all but the last `*' in a group of consecutive
		 * `*' characters to avoid unnecessary glob2() recursion.
		 */
		while (*p++ == ASTERISK)
			;	/* nothing */
		if (*--p == EOS)
			return true;
		e--;
		while (*e != EOS)
			if (glob2(e++, p))
				return true;
		break;

	case QUESTION:
		if (ec != EOS)
			return glob2(e, p);
		break;

	case LBRACKET:
		if (*p == EOS)
			break;
		for (c = UCHAR(EOS), cok = rok = 0, n = p + 1; ; ) {
			pc = *p++;
			if (pc == RBRACKET && p > n) {
				if (cok > 0 || rok > 0)
					return glob2(e, p);
				break;
			}
			if (*p == EOS)
				break;
			if (pc == HYPHEN && c != EOS && *p != RBRACKET) {
				pc = *p++ & ASCII;
				if (*p == EOS)
					break;
				if (c <= ec && ec <= pc)
					rok++;
				else if (c == ec)
					cok--;
				c = UCHAR(EOS);
			} else {
				c = pc & ASCII;
				if (ec == c)
					cok++;
			}
		}
		break;

	default:
		if ((pc & ASCII) == ec)
			return glob2(e, p);
	}
	return false;
}

static void
gsort(const char **ogavp)
{
	const char **p1, **p2, *sap;

	p1 = ogavp;
	while (gavp - p1 > 1) {
		p2 = p1;
		while (++p2 < gavp)
			if (strcmp(*p1, *p2) > 0) {
				sap = *p1;
				*p1 = *p2;
				*p2 = sap;
			}
		p1++;
	}
}

static DIR *
gopendir(const char *dname)
{
	char *b, buf[PATHMAX];
	const char *d;

	for (*buf = EOS, b = buf, d = dname; b < &buf[PATHMAX]; b++, d++)
		if ((*b = (*d & ASCII)) == EOS)
			break;
	return (b >= &buf[PATHMAX]) ? NULL : opendir(buf);
}
